File: 0838-design-linked-list/0838-design-linked-list.py
class ListNode:
    def __init__(self, val):
        self.val = val
        self.next = None

class MyLinkedList(object):

    def __init__(self):
        self.head = None
        self.tail = None
        self.size = 0

    def get(self, index):
        """
        :type index: int
        :rtype: int
        """

        if((self.size == 0) or (index > self.size) or (index < 0) or (index == self.size) ):
            return -1
        
        if index == 0:
            return self.head.val
        
        if index == (self.size - 1):
            return self.tail.val
        
        
        desired_index = 0

        dummy = self.head

        while desired_index != index:

            dummy = dummy.next
            desired_index += 1
        
        return dummy.val


        

    def addAtHead(self, val):
        """
        :type val: int
        :rtype: None
        """
        if self.head == None and self.tail == None:
            nextUp = ListNode(val)

            self.head = self.tail = nextUp
            self.size += 1
        else:
            nextUp = ListNode(val)
           
            nextUp.next = self.head
            self.head = nextUp
            self.size += 1
 
        

    def addAtTail(self, val):
        """
        :type val: int
        :rtype: None
        """
        if self.head == None and self.tail == None:
            nextUp = ListNode(val)

            self.head = self.tail = nextUp
            self.size += 1
        else:
            nextUp = ListNode(val)
            self.tail.next = nextUp
            self.tail  = self.tail.next 
            self.size += 1
 

    def addAtIndex(self, index, val):
        """
        :type index: int
        :type val: int
        :rtype: None
        """
        if(index > self.size or index < 0):
            return
        elif(index == self.size):
            self.addAtTail(val)
        elif(index == 0):
            self.addAtHead(val)
        else:
            
            desired_index = 0

            dummy = self.head

            while desired_index !=  index - 1:

                dummy = dummy.next
                desired_index += 1
            
            nextUp = ListNode(val)

            nextUp.next = dummy.next
            dummy.next = nextUp
            self.size += 1
        

    def deleteAtIndex(self, index):
        """
        :type index: int
        :rtype: None
        """
        print(index)
        print(self.size)

        if(index > self.size or index < 0 or index == self.size):
            return
        
        if index == 0 and self.head == self.tail and self.size == 1:
            self.size -= 1
            self.head = self.tail = None
        elif self.size == 2:
            if index == 0:
                self.head = self.tail
                self.size -= 1
            elif index == 1:
                self.head.next = None
                self.tail = self.head
                self.size -= 1
        else:
            if index == 0:
                self.head = self.head.next
                self.size -= 1
                return
            
            desired_index = 0

            dummy = self.head

            while desired_index !=  index - 1:

                dummy = dummy.next
                desired_index += 1
            
            if(index == (self.size - 1)):
                dummy.next = None
                self.tail = dummy
            else:
                dummy.next = dummy.next.next
            self.size -= 1
        




# Your MyLinkedList object will be instantiated and called as such:
# obj = MyLinkedList()
# param_1 = obj.get(index)
# obj.addAtHead(val)
# obj.addAtTail(val)
# obj.addAtIndex(index,val)
# obj.deleteAtIndex(index)

File: 0838-design-linked-list/0838-design-linked-list.js
var Node = function(value) {
    this.value = value;
    this.next = null;
    this.prev = null;
};


var MyLinkedList = function() {
    this.head = null;  // Top of the stack
    this.last = null;   // Bottom of the stack
    this.length = 0;
};

/** 
 * @param {number} index
 * @return {number}
 */
MyLinkedList.prototype.get = function(index) {
    
    if(index >= 0 && index < this.length){
        let dummy = this.head;
        for(let i = 0; i < this.length; i++){
            if(i === index){
                return dummy.value;
            }else{
                if(dummy)  {
                    console.log("hi", dummy.value)
                    dummy = dummy.next;
                }
               
            }
        }
    }

    return -1;
};

/** 
 * @param {number} val
 * @return {void}
 */
MyLinkedList.prototype.addAtHead = function(val) {
    this.addAtIndex(0, val);
};

/** 
 * @param {number} val
 * @return {void}
 */
MyLinkedList.prototype.addAtTail = function(val) {
    this.addAtIndex(this.length, val);
};

/** 
 * @param {number} index 
 * @param {number} val
 * @return {void}
 */
MyLinkedList.prototype.addAtIndex = function(index, val) {

    if(index >= 0 && index <= this.length){
        if(index === 0){
            let newNode = new Node(val);
            if(this.length === 0){
                this.head = newNode;
                this.last = this.head;
            }else{
                newNode.next = this.head;
                this.head.prev = newNode;
                this.head = newNode;
            }

        } else if(index === this.length){
            let newNode = new Node(val);
            newNode.prev = this.last;
            this.last.next = newNode;
            this.last = newNode;
        } else{
            let dummy = this.head;
            for(let i = 0; i < this.length; i++){
                if((i+1) === index){
                    let newNode = new Node(val);
                    newNode.prev = dummy;
                    newNode.next = dummy.next;
                    dummy.next.prev = newNode;
                    dummy.next = newNode;
                    break;
                }else{
                    dummy = dummy.next;
                }
            }
        }
        this.length += 1;
    }
};

/** 
 * @param {number} index
 * @return {void}
 */
MyLinkedList.prototype.deleteAtIndex = function(index) {

   
    if(index >= 0 && index < this.length){
        if(index === 0){
            this.head = this.head.next;
        } else if(index === this.length - 1){
            this.last = this.last.prev;
            this.last.next = null;
        }else{
            let dummy = this.head;
            
            for(let i = 0; i < this.length; i++){
                if(i === index){
                    let dummiesNext = dummy.next;
                    let dummiesPrev = dummy.prev;
                    dummiesNext.prev = dummiesPrev;
                    dummiesPrev.next = dummiesNext;
                    break;
                }else{
                    dummy = dummy.next;
                }
            }
        }
        this.length -= 1;
    }
    
};

/** 
 * Your MyLinkedList object will be instantiated and called as such:
 * var obj = new MyLinkedList()
 * var param_1 = obj.get(index)
 * obj.addAtHead(val)
 * obj.addAtTail(val)
 * obj.addAtIndex(index,val)
 * obj.deleteAtIndex(index)
 */